### PR TITLE
Do not rebuffer socket streams for web sockets, other fixes.

### LINKED ISSUE #1071: BufferedSource error with WebSockets
In `WebSocketListener#onMessage(BufferedSource bufferedSource, WebSocket.PayloadType payloadType)`, calling `bufferedSource.readUtf8()` after verifying that `payloadType == WebSocket.PayloadType.TEXT` throws the following exception:

```
System.err  W  java.lang.IllegalArgumentException: byteCount < 0: -26991
    W      at okio.RealBufferedSource.read(RealBufferedSource.java:46)
    W      at com.squareup.okhttp.internal.ws.WebSocketReader$FramedMessageSource.read(WebSocketReader.java:250)
    W      at okio.RealBufferedSource.exhausted(RealBufferedSource.java:60)
    W      at com.operator.expert.service.ChatSocketConnection.onMessage(ChatSocketConnection.java:76)
    W      at com.squareup.okhttp.internal.ws.WebSocketReader.readMessage(WebSocketReader.java:112)
    W      at com.squareup.okhttp.WebSocket$ReaderRunnable.execute(WebSocket.java:283)
    W      at com.squareup.okhttp.internal.NamedRunnable.run(NamedRunnable.java:33)
    W      at java.lang.Thread.run(Thread.java:841)
```

Any ideas how to prevent/fix this? It doesn't appear to happen to all socket messages.

### LINKED ISSUE #1296: WebSocketCallTest.serverMessage is flaky
```
<<< FAILURE! - in com.squareup.okhttp.internal.ws.WebSocketCallTest
serverMessage(com.squareup.okhttp.internal.ws.WebSocketCallTest)  Time elapsed: 10.014 sec  <<< FAILURE!
java.lang.AssertionError: expected:<Message[TEXT Buffer[size=18 md5=be471905fcbe7f881ceb7fa60775ab5e]]> but was:<null>
	at org.junit.Assert.fail(Assert.java:88)
	at org.junit.Assert.failNotEquals(Assert.java:743)
	at org.junit.Assert.assertEquals(Assert.java:118)
	at org.junit.Assert.assertEquals(Assert.java:144)
	at com.squareup.okhttp.internal.ws.WebSocketRecorder.assertTextMessage(WebSocketRecorder.java:89)
	at com.squareup.okhttp.internal.ws.WebSocketCallTest.serverMessage(WebSocketCallTest.java:78)
```

### PR DESCRIPTION
**Pull request by JakeWharton**
* Ensure onOpen is called before the reader thread starts.
* Allow close reason without code.
* Use ProtocolException for protocol errors to ensure a close response is written.

Closes #1071. Closes #1296.

### CODE REVIEW

**Comment by JakeWharton** [comment]
This does not bring us into full compliance, but it moves us a lot closer.

File: okhttp-ws-tests/src/test/java/com/squareup/okhttp/ws/AutobahnTester.java [hunk 1]
```diff
@@ -0,0 +1,159 @@
+package com.squareup.okhttp.ws;
```
Comment (swankjesse) [inline okhttp-ws-tests/src/test/java/com/squareup/okhttp/ws/AutobahnTester.java]: copyright

File: okhttp-ws-tests/src/test/java/com/squareup/okhttp/ws/AutobahnTester.java [hunk 2]
```diff
@@ -0,0 +1,159 @@
...
+      }
+
+      updateReports();
+    } finally {
+      client.getDispatcher().getExecutorService().shutdown();
+    }
+  }
+
+  private void runTest(final long number, final long count) throws IOException {
+    final AtomicReference<WebSocket> webSocketRef = new AtomicReference<>();
```
Comment (swankjesse) [inline okhttp-ws-tests/src/test/java/com/squareup/okhttp/ws/AutobahnTester.java]: Could this just be a member of the `WebSocketListener` anonymous class?

File: okhttp-ws-tests/src/test/java/com/squareup/okhttp/ws/AutobahnTester.java [hunk 3]
```diff
@@ -0,0 +1,159 @@
...
+          @Override public void onFailure(IOException e) {
+            latch.countDown();
+          }
+        });
+    try {
+      if (!latch.await(10, TimeUnit.SECONDS)) {
+        throw new IllegalStateException("Timed out waiting for count.");
+      }
+    } catch (InterruptedException e) {
+      Thread.currentThread().interrupt();
```
Comment (swankjesse) [inline okhttp-ws-tests/src/test/java/com/squareup/okhttp/ws/AutobahnTester.java]: Is interruption expected? I'd be tempted to just throw an assertion error here (and not presume to support interruption at all)

File: okhttp-ws-tests/src/test/java/com/squareup/okhttp/ws/AutobahnTester.java [hunk 4]
```diff
@@ -0,0 +1,159 @@
...
+    final AtomicLong countRef = new AtomicLong();
+    final AtomicReference<IOException> failureRef = new AtomicReference<>();
+    newWebSocket("/getCaseCount").enqueue(new WebSocketListener() {
+      @Override public void onOpen(WebSocket webSocket, Request request, Response response)
+          throws IOException {
+      }
+
+      @Override public void onMessage(BufferedSource payload, WebSocket.PayloadType type)
+          throws IOException {
+        countRef.set(payload.readDecimalLong());
```
Comment (swankjesse) [inline okhttp-ws-tests/src/test/java/com/squareup/okhttp/ws/AutobahnTester.java]: readDecimalLong FTW

**Comment by swankjesse** [comment]
LGTM

File: okhttp-ws-tests/src/test/java/com/squareup/okhttp/ws/AutobahnTester.java (hunk 2 above)
Comment (JakeWharton) [inline okhttp-ws-tests/src/test/java/com/squareup/okhttp/ws/AutobahnTester.java]: Yep

**Pull request merged**